crates/sparrow-runtime/src/execute/output/object_store.rs

```diff
@@ -1,209 +1,263 @@
-
-        let tmp_output_file_name = output_file_name(format)?;
-        let tmp_output_file = create_tempfile(&tmp_output_file_name)?;
-        let tmp_output_path = tmp_output_file.path().to_owned();
-
-        // Write the results to the specified format
-        write_to_file(
-            tmp_output_file.as_file(),
-            format,
-            schema,
-            progress_updates_tx.clone(),
-            batches,
-        )
-        .await?;
+        .change_context(Error::ProgressUpdate)?;
 
-        // Upload the temp local file to S3
-        let mut remote_output_file = S3Object::try_from_uri(&output_prefix)
-            .into_report()
-            .change_context(Error::MalformedUri)?;
-        remote_output_file.push_delimited(&tmp_output_file_name);
-        let output_path = remote_output_file.get_formatted_key();
+    let output_prefix = ObjectStoreUrl::from_str(&destination.output_prefix_uri)
+        .change_context(Error::InvalidDestination)?;
+    let object_store = object_stores
+        .object_store(&output_prefix)
+        .change_context(Error::InvalidDestination)?;
 
-        s3_client
-            .upload_s3(remote_output_file.clone(), &tmp_output_path)
-            .await
-            .into_report()
-            .change_context(Error::UploadFailure)?;
+    let mut num_rows_in_file = 0;
 
-        progress_updates_tx
-            .try_send(ProgressUpdate::FilesProduced {
-                paths: vec![output_path],
-            })
-            .into_report()
-            .change_context(Error::ProgressUpdateFailure)?
-    } else {
-        // Assumes local storage otherwise now
-        let output_prefix = output_prefix
-            .strip_prefix("file://")
-            .ok_or(Error::UnsupportedUri)
-            .into_report()
-            .attach_printable_lazy(|| {
-                format!("unsupported output prefix uri {output_prefix}, expected local file (file:///...)")
-            })?;
-
-        if !output_prefix.starts_with('/') {
-            return Err(
-                error_stack::report!(Error::UnsupportedUri).attach_printable(format!(
-                    "output prefix uri requires absolute path, got {output_prefix}"
-                )),
+    // Always create at least one writer. This ensures that if there are
+    // no rows, we still emit an empty file.
+    let mut state = Some(
+        WriterState::open(
+            object_store.as_ref(),
+            &output_prefix,
+            destination.file_type(),
+            schema.clone(),
+        )
+        .await?,
+    );
+
+    // Currently, we pull batches and upload them asynchronously.
+    // We could increase concurrency by buffering some batches and performing
+    // multiple writes simultaneously.
+    while let Some(batch) = batches.next().await {
+        if batch.num_rows() == 0 {
+            // Don't create empty files. Wait for a non-empty batch before
+            // starting a new file.
+            continue;
```

Comment:
If we have no output, is the initial empty file with headers still created?

Reply:
Yes. I originally switched to having no files, but that causes problems for tests. It seems better to always ensure we have at least one output file, even if it has no rows. So, this is why (as noted in the comment above) we create the initial state as `Some(...)` instead of starting with `None`. In the `WriterState::open` we also write an empty batch so that the header is added to the CSV, etc.